File: 13-Graph-Algorithms/09-Graph-With-DP/java/LongestPathDAG.java
import java.util.*;
public class LongestPathDAG {
    public static void main(String[] args) {
        int n = 6; List<int[]>[] adj = new List[n];
        for (int i=0;i<n;i++) adj[i]=new ArrayList<>();
        adj[0].add(new int[]{1,5}); adj[0].add(new int[]{2,3}); adj[1].add(new int[]{3,6});
        adj[1].add(new int[]{2,2}); adj[2].add(new int[]{4,4}); adj[2].add(new int[]{5,2});
        int[] indeg=new int[n]; for(int u=0;u<n;u++) for(int[] e:adj[u]) indeg[e[0]]++;
        Queue<Integer> q=new LinkedList<>(); for(int i=0;i<n;i++) if(indeg[i]==0) q.add(i);
        List<Integer> topo=new ArrayList<>(); while(!q.isEmpty()){int u=q.poll();topo.add(u);for(int[] e:adj[u])if(--indeg[e[0]]==0)q.add(e[0]);}
        int[] dist=new int[n]; Arrays.fill(dist,Integer.MIN_VALUE); dist[0]=0;
        for(int u:topo) if(dist[u]!=Integer.MIN_VALUE) for(int[] e:adj[u]) dist[e[0]]=Math.max(dist[e[0]],dist[u]+e[1]);
        for(int i=0;i<n;i++) System.out.println(i+": "+dist[i]);
    }
}

File: 13-Graph-Algorithms/09-Graph-With-DP/cpp/longest_path_dag.cpp
/* Graph + DP: Longest Path in DAG | Time: O(V+E) | Tags: graph-dp, dag, topological */
#include <bits/stdc++.h>
using namespace std;
int main(){
    int n=6;
    vector<vector<pair<int,int>>> adj(n);
    // DAG edges: u->v with weight w
    adj[0].push_back({1,5}); adj[0].push_back({2,3}); adj[1].push_back({3,6}); adj[1].push_back({2,2});
    adj[2].push_back({4,4}); adj[2].push_back({5,2}); adj[3].push_back({5,1}); adj[3].push_back({4,-1});
    adj[4].push_back({5,1});
    
    // Topological sort
    vector<int> indeg(n,0); for(int u=0;u<n;u++) for(auto&[v,w]:adj[u]) indeg[v]++;
    queue<int> q; for(int i=0;i<n;i++) if(!indeg[i]) q.push(i);
    vector<int> topo; while(!q.empty()){int u=q.front();q.pop();topo.push_back(u);for(auto&[v,w]:adj[u])if(--indeg[v]==0)q.push(v);}
    
    // DP: longest path from source 0
    vector<int> dist(n, INT_MIN); dist[0]=0;
    for(int u:topo) if(dist[u]!=INT_MIN) for(auto&[v,w]:adj[u]) dist[v]=max(dist[v],dist[u]+w);
    
    cout<<"Longest distances from 0:"<<endl;
    for(int i=0;i<n;i++) cout<<i<<": "<<(dist[i]==INT_MIN?"INF":to_string(dist[i]))<<endl;
}
